src/coreclr/jit/copyprop.cpp

```diff
@@ -190,7 +202,129 @@ bool Compiler::optCopyProp(
+                                        treePhiArgSsaDefVN);
+                                treePhiArg->SetVNs(treePhiArgSsaDef->m_vnPair);
+                            }
+
+                            treePhiArgVN = treePhiArgSsaDefVN;
+                        }
+
+                        if ((otherPhiArgVN == ValueNumStore::NoVN) && phiArgIsFromDfsBackedge)
+                        {
+                            LclSsaVarDsc* const otherPhiArgSsaDef =
+                                lvaGetDesc(otherPhiArg)->GetPerSsaData(otherPhiArg->GetSsaNum());
+                            ValueNum otherPhiArgSsaDefVN = otherPhiArgSsaDef->m_vnPair.GetConservative();
+
+                            // Opportunisitcally update the PhiArgVNs, both to short-circuit future checks here
+                            // in copy prop, and help later phases that also scan phi arg lists (eg RBO).
+                            // (consider doing this as a VN post pass?)
+                            //
+                            if (otherPhiArgSsaDefVN != otherPhiArgVN)
+                            {
+                                JITDUMP("Updating backedge phi arg [%06u] vn to " FMT_VN "\n", dspTreeID(otherPhiArg),
+                                        otherPhiArgSsaDefVN);
+                                otherPhiArg->SetVNs(otherPhiArgSsaDef->m_vnPair);
+                            }
+
+                            otherPhiArgVN = otherPhiArgSsaDefVN;
+                        }
+
+                        // If the updated VNs differ, the phis are not equivalent
+                        //
+                        if (treePhiArgVN != otherPhiArgVN)
+                        {
+                            phiArgsAreEquivalent = false;
+                            break;
+                        }
+
+                        // If we failed to find meaningful VNs, the phis are not equivalent
+                        //
+                        if (treePhiArgVN == ValueNumStore::NoVN)
+                        {
+                            phiArgsAreEquivalent = false;
+                            break;
+                        }
+                    }
+
+                    // If we didn't verify all phi args we have failed to prove equivalence
+                    //
+                    phiArgsAreEquivalent &= (treeIter == treeEnd);
+                    phiArgsAreEquivalent &= (otherIter == otherEnd);
+
+                    if (phiArgsAreEquivalent)
+                    {
+                        JITDUMP("Found equivalent phi defs: [%06u] and [%06u]\n", dspTreeID(treePhi),
+                                dspTreeID(otherPhi));
+                        foundEquivalentPhi = true;
+                    }
+                }
+            }
+
+            // VNs differed, and we didn't find phis equivalent, move on to the next candidate.
+            //
+            if (!foundEquivalentPhi)
+            {
+                continue;
+            }
```

Comment:
The nesting here is pretty deep... consider introducing a function with some early outs?

Reply:
Yeah, I was tempted to try that to see if helped TP any, but haven't yet.